### PR TITLE
WIP: kvutils batch validator

### PR DESCRIPTION
**Pull request by dajmaki**
Trying out parallel validation of a batch. Still very much work in progress.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateWriter.scala [hunk 1]
```diff
@@ -6,20 +6,34 @@ package com.daml.ledger.participant.state.kvutils.api
...
     extends WriteService {
 
+  implicit val executionContext: ExecutionContext = materializer.executionContext
+
+  val batchingWriter = new BatchingLedgerWriter(
+    writer,
+    1024,
+    4 * 1024 * 1024,
+    100.millis,
+    10
```
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateWriter.scala]: Can we use named arguments here for clarity?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/BatchValidator.scala [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+import scala.util.Try
+
+// This is example code for validating a batch of submissions in parallel, including conflict detection.
+class BatchValidator[LogResult](
+    val allocateLogEntryId: () => DamlLogEntryId,
+    val ledgerStateAccess: LedgerStateAccess[LogResult]) {
+
+  // Queue for validation work.
+  val validationQueue = new ArrayBlockingQueue[Runnable](1024)
+  val nThreads = 2 * Runtime.getRuntime.availableProcessors()
```
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/BatchValidator.scala]: I like constants like this to be parameters in the constructor, with a default argument. Makes it clear that it's configuration.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/BatchValidator.scala [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+  // Executor that performs the validations.
+  // TODO(JM): We probably don't need anything this complicated. A fixed thread pool should be fine.
+  // Or if we want to keep things simpler we can just use an existing thread pool / execution context,
+  // though it would be useful for profiling purposes to have a dedicated thread pool.
+  val executor = new ThreadPoolExecutor(
```
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/BatchValidator.scala]: Nothing cleans this up. Can we make this `Closeable`, or better yet, injected?

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/api/BatchingLedgerWriterSpec.scala [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      res3 <- batchingWriter.commit("test3", Array[Byte](11, 12, 13, 14, 15))
+    } yield {
+      verify(mockWriter).commit(ArgumentMatchers.eq("test1"), ArgumentMatchers.any[Array[Byte]])
+      verify(mockWriter).commit(ArgumentMatchers.eq("test2"), ArgumentMatchers.any[Array[Byte]])
+      verify(mockWriter).commit(ArgumentMatchers.eq("test3"), ArgumentMatchers.any[Array[Byte]])
+
+      all(Seq(res1, res2, res3)) should be(SubmissionResult.Acknowledged)
+    }
+  }
+}
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/api/BatchingLedgerWriterSpec.scala]: Same as https://github.com/digital-asset/daml/pull/4964 I think, please refer to comments there

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala [hunk 5]
```diff
@@ -74,6 +74,20 @@ class SubmissionValidator[LogResult](
...
+  private val batchValidator = new BatchValidator[LogResult](allocateLogEntryId, ledgerStateAccess)
+  private[validator] def validateAndCommitWithBatchValidator(
+      envelope: RawBytes,
+      correlationId: String,
+      recordTime: Timestamp,
+      participantId: ParticipantId,
+  )(implicit logCtx: LoggingContext): Future[Either[ValidationFailed, LogResult]] =
+    batchValidator.validate(recordTime, participantId, envelope).map { results =>
+      // FIXME(JM): HACK
+      Right(results.last)
```
[the remark this answers is not in the document]
Comment (dajmaki) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala]: Completely implementation dependent, e.g. the in-memory one uses it to signal new head index. We'll need to expose multiple results to the caller.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/BatchValidator.scala (hunk 3 above)
Comment (dajmaki) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/BatchValidator.scala]: What do you mean injected?

Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/BatchValidator.scala]: Passed in through the constructor. Basically, make it Somebody Else's Problem™.

Of course, you're also in charge of the Somebody Else. Making a `ResourceOwner` might help. See `DbDispatcher` for an example.

**Pull request closed without merging**